ding/model/template/qtransformer.py

```diff
@@ -0,0 +1,753 @@
+from random import random
+from functools import partial, cache
```

Comment:
`cache` is the new feature in python3.9, for compatibility, you should implement it as follows:
```python
try:
    from functools import cache  # only in Python >= 3.9
except ImportError:
    from functools import lru_cache
    cache = lru_cache(maxsize=None)
```